java/arcs/core/analysis/RecipeGraphFixpointIterator.kt

```diff
@@ -0,0 +1,107 @@
+
+    /** State transformer for a [Recipe.Handle] node. */
+    open fun transform(handle: Recipe.Handle, input: V): V = input
+
+    /** State transformer for a [Recipe.Particle] node. */
+    open fun transform(particle: Recipe.Particle, input: V): V = input
+
+    /** State transformer for a [Recipe.Handle] -> [Recipe.Particle] edge. */
+    open fun transform(
+        handle: Recipe.Handle,
+        particle: Recipe.Particle,
+        spec: HandleConnectionSpec,
+        input: V
+    ): V = input
+
+    /** State transformer for a [Recipe.Particle] -> [Recipe.Handle] edge. */
+    open fun transform(
+        particle: Recipe.Particle,
+        handle: Recipe.Handle,
+        spec: HandleConnectionSpec,
+        input: V
+    ): V = input
+
+    /**
+     * Returns the initial values for the nodes for starting a fixpoint iteration.
+     *
+     * The nodes in the returned map are used to initialize the worklist for the fixpoint iteration.
+     */
+    abstract fun getInitialValues(graph: RecipeGraph): Map<RecipeGraph.Node, V>
+
+    /** Returns the value for the given particle. Returns null if the value is BOTTOM. */
+    fun getValue(particle: Recipe.Particle): V? = nodeValues[RecipeGraph.Node.Particle(particle)]
+
+    /** Returns the value for the given handle. Returns null if the value is BOTTOM. */
+    fun getValue(handle: Recipe.Handle): V? = nodeValues[RecipeGraph.Node.Handle(handle)]
+
+    fun computeFixpoint(graph: RecipeGraph) {
+        // TODO(bgogul): If there are identical particles in the recipe, the results will be messed
+        // up. Need to fix the issue.
+        nodeValues = getInitialValues(graph).toMutableMap()
+        val worklist = nodeValues.keys.toMutableSet()
+        while (worklist.isNotEmpty()) {
+            // Pick and remove an element from the worklist.
+            val current = worklist.first()
+            worklist.remove(current)
+            val input = nodeValues[current]
+            // Treating as BOTTOM now. See `successors.forEach` below as well.
+            if (input == null) continue
+            val output = transform(current, input)
+            current.successors.forEach { (succ, spec) ->
+                val edgeValue = transform(current, succ, spec, output)
+                val oldValue = nodeValues[succ]
+                val newValue = oldValue?.join(edgeValue) ?: edgeValue
+                val changed = oldValue?.isEquivalentTo(newValue)?.not() ?: true
+                if (changed) {
+                    worklist.add(succ)
+                    nodeValues[succ] = newValue
+                }
+            }
+        }
+    }
+
+    private fun transform(node: RecipeGraph.Node, input: V) = when (node) {
+        is RecipeGraph.Node.Particle -> transform(node.particle, input)
+        is RecipeGraph.Node.Handle -> transform(node.handle, input)
+    }
+
+    private fun transform(
+        src: RecipeGraph.Node,
+        tgt: RecipeGraph.Node,
```

Comment:
`tgt` is a bit obscure. How about `source` and `target`? Or `start` and `end`?

Reply:
Done.